x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java

```diff
@@ -166,24 +153,79 @@ protected void doExecute(Task task, GetDataFrameAnalyticsStatsAction.Request req
-            allTasksStats, allTasksStats.size(), GetDataFrameAnalyticsAction.Response.RESULTS_FIELD)));
+        if (stoppedTasksIds.isEmpty()) {
+            listener.onResponse(runningTasksResponse);
+            return;
+        }
+
+        searchStoredProgresses(stoppedTasksIds, ActionListener.wrap(
+            storedProgresses -> {
+                List<Stats> stoppedStats = new ArrayList<>(stoppedTasksIds.size());
+                for (int i = 0; i < stoppedTasksIds.size(); i++) {
+                    String configId = stoppedTasksIds.get(i);
+                    StoredProgress storedProgress = storedProgresses.get(i);
+                    stoppedStats.add(buildStats(configId, storedProgress.get()));
+                }
+                List<Stats> allTasksStats = new ArrayList<>(runningTasksResponse.getResponse().results());
+                allTasksStats.addAll(stoppedStats);
+                Collections.sort(allTasksStats, Comparator.comparing(Stats::getId));
+                listener.onResponse(new GetDataFrameAnalyticsStatsAction.Response(new QueryPage<>(
+                    allTasksStats, allTasksStats.size(), GetDataFrameAnalyticsAction.Response.RESULTS_FIELD)));
+            },
+            listener::onFailure
+        ));
     }
 
     static List<String> determineStoppedTasksIds(List<String> expandedIds, List<Stats> runningTasksStats) {
         Set<String> startedTasksIds = runningTasksStats.stream().map(Stats::getId).collect(Collectors.toSet());
         return expandedIds.stream().filter(id -> startedTasksIds.contains(id) == false).collect(Collectors.toList());
     }
 
-    private GetDataFrameAnalyticsStatsAction.Response.Stats buildStatsForStoppedTask(String concreteAnalyticsId) {
-        return buildStats(concreteAnalyticsId, null);
+    private void searchStoredProgresses(List<String> configIds, ActionListener<List<StoredProgress>> listener) {
+        MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+        for (String configId : configIds) {
+            SearchRequest searchRequest = new SearchRequest(AnomalyDetectorsIndex.jobStateIndexPattern());
+            searchRequest.indicesOptions(IndicesOptions.lenientExpandOpen());
+            searchRequest.source().size(1);
+            searchRequest.source().query(QueryBuilders.idsQuery().addIds(DataFrameAnalyticsTask.progressDocId(configId)));
+            multiSearchRequest.add(searchRequest);
+        }
+
+        executeAsyncWithOrigin(client, ML_ORIGIN, MultiSearchAction.INSTANCE, multiSearchRequest, ActionListener.wrap(
+            multiSearchResponse -> {
+                List<StoredProgress> progresses = new ArrayList<>(configIds.size());
+                for (MultiSearchResponse.Item itemResponse : multiSearchResponse.getResponses()) {
+                    if (itemResponse.isFailure()) {
+                        listener.onFailure(ExceptionsHelper.serverError(itemResponse.getFailureMessage(), itemResponse.getFailure()));
```

Comment:
```suggestion
                        listener.onFailure(ExceptionsHelper.serverError(itemResponse.getFailureMessage(), itemResponse.getFailure()));
                        return;
```

I think if you don't return here, and you have a failure, there will be issues with the same listener being called twice.

Reply:
Classic! Thanks for watching over my shoulder :-)